### ISSUE TITLE
Utils::HeaderHash is EXTREMELY expensive

### ISSUE THREAD

**Issue by schneems** [rack/rack#738]
Hypothetically if it was safe to remove HeaderHash usage right now, we would see

```
rack/lib x 7318
rack/lib x 5678
```

22.4 % fewer objects per request, and here's the kicker:

Speed with Header Hash

```
Calculating -------------------------------------
                 ips        51 i/100ms
-------------------------------------------------
                 ips      524.8 (±9.3%) i/s -       2601 in   5.005893s
```

Speed Without header Hash

```
Calculating -------------------------------------
                 ips        62 i/100ms
-------------------------------------------------
                 ips      649.2 (±8.8%) i/s -       3224 in   5.005930s
```

We get a `(649.2 - 524.8 )/524.8  * 100 # => 23` % faster response only by removing HeaderHash. This is against a whole Rails app, not just Rack benchmarks in isolation. I don't know about you, but that sounds like a lot.
## Reality

I know we can't remove `HeaderHash` wholesale, but we should do an audit of the places it's used, maybe it's not necessarry everywhere. 

Perhaps we can cross reference documentation and for all classes that explicitly mention returning a header hash with this behavior (preserves original keys, and but allows for access with varying case), if documentation doesn't exist re-write to not use HeaderHash. Is that crazy? This util class is new to me, so I would appreciate any background info you might have.

With no changes or fixes, wholesale removing HeaderHash behavior from Rack causes 5 failures out of 2020 assertions in the test suite.

**Comment by wjordan** [comment]
@schneems do you have a working repo or other steps to replicate your benchmarking tests?

I did a bit of exploration and it looks possible to reduce the number of excessive per-request object allocations caused by HeaderHash by memoizing the results of `downcase` methods into a class variable, e.g.: f17d66dab152da4e5eb5ad427c88b66af57442b9

If the cost of using HeaderHash is mainly due to string object allocations caused by `downcase`, then a change like this could be worth the slight extra cost of extra memory usage, and an internal performance fix could be more easily backported without breaking API changes.

Using this technique I see a slight reduction in total object allocations (from 1173 to 1131 objects profiling a simple 'hello world' Rails 4.2 app with [derailed](https://github.com/schneems/derailed_benchmarks)), but negligible `ips` improvement w/Ruby 2.2- perhaps this is no longer an issue with the updated Ruby GC? Or do some Rails apps use HeaderHash much more extensively than the default `rails new` stack I tested on?

**Comment by raggi** [comment]
Prior experiment FTR: https://github.com/methodmissing/hwia

**Comment by raggi** [comment]
Someone should make a case-insensitive hash for ruby-core. I mean, ruby desperately needs efficient trees, dicts, all kinds of data structures are missing that can never be efficiently implemented at a high level due to the weight of RObject.

**Comment by schneems** [comment]
@raggi proposals for 2.3 are being discussed now. I know that they're not keen on expanding the stdlib since it is a maintenance burden and theoretically such functionality could be written into a gem. If you have some ideas on things that would make maintainers lives easier/faster now would be the time. Alternatively we could deprecate the case insensitive behavior and phase it out in Rack 2.0. WDYT?

@wjordan I used derailed_benchmarks and codetriage, or maybe it was a small rack app with RackMock and benchmark/ips. I think I just replaced HeaderHash with a regular hash, and benchmarked, I didn't dig into exactly what was expensive.

**Comment by raggi** [comment]
@schneems If someone were to write a portable gem for this then it might be a reasonable dependency for Rack. That's @tenderlove 's call now really. You could clean up hwia, add JRuby support and pure ruby fallback, and release native versions of the gems, etc. It's probably more work getting a good release setup going than it is to write the thing in the first place.

Another very real option (although not necessarily a good one), is to forcefully downcase everything. http2 effectively does this anyway. It could break some very old things.

You could do that and make it configurable. I'm sure there are more options if we brainstorm it.

**Comment by jkowens** [comment]
It looks like @jodosha has identified a performance improvement for `HeaderHash` in #1325. Maybe that will resolve this issue if merged?

**Comment by ioquatix** [comment]
I know this is a tricky issue but here are two points of reference:

Protocol specific headers in pure Ruby. Probably not as fast as it could be.

https://github.com/socketry/http-protocol/blob/master/lib/http/protocol/headers.rb

A C++ implementation of case-insensitive headers:

https://github.com/kurocha/async-http/blob/master/source/Async/HTTP/Headers.hpp

It uses case insensitive compare and hash functions:

https://github.com/kurocha/async-http/blob/master/source/Async/HTTP/Headers.cpp

I think one of the main issues that I tried to address with the pure Ruby implementation is that while people generally think of headers as a hash table or key-value pairs, it's actually a lot more complex than that.

**Comment by jeremyevans** [comment]
Do we still want to do anything about this?  #1325 was merged and improved performance somewhat, but I'm not sure how much.

One possible approach that isn't perfect but may work for the middleware:

* Remove use of HeaderHash where it is currently used in middleware
* Replace `[]` and `delete` calls on the header hash with `Utils` functions that try the following keys in order:
  * Lookup Expected capitalization (`Content-Length`)
  * Lookup Lowercase (`content-length`)
  * Lookup Uppercase (`CONTENT-LENGTH`)

This should be very little performance difference in the common case of the key existing with expected capitalization.  It's probably a decent performance hit for keys that don't exist, but still likely to be much less than creating the HeaderHash in each middleware.  This will miss keys with weird capitalization (`CoNtEnT-LeNgTh`), but I'm not sure whether that matters.  As these are response headers, they are assumingly generated by the application and not attacker controlled.

I don't think it's possible to use this approach for Response, since that API allows setting arbitrary headers.  Still I think the above approach could significantly improve performance of the middleware.

Here's a benchmark for `Rack::ContentLength`:

```ruby
require 'benchmark/ips'
require 'rack'

app = lambda{|_| [200, {'Content-Length'=>'1'}, ['a']]}
wrapp = Rack::ContentLength.new(app)

app2 = lambda{|_| [200, {}, ['a']]}
wrapp2 = Rack::ContentLength.new(app)

h = {}
Benchmark.ips do |x|
  x.report("direct"){app.call(h)}
  x.report("Rack::Content-Length"){wrapp.call(h)}
  x.report("direct-miss"){app2.call(h)}
  x.report("Rack::Content-Length-miss"){wrapp2.call(h)}
end
```

Results using HeadersHash (current rack master 8eb03292408f3bf6093ede1858159c616a9daadc):

```
Warming up --------------------------------------
              direct    46.084k i/100ms
Rack::Content-Length    11.941k i/100ms
         direct-miss    58.702k i/100ms
Rack::Content-Length-miss
                        11.872k i/100ms
Calculating -------------------------------------
              direct    920.401k (_ 0.6%) i/s -      4.608M in   5.007151s
Rack::Content-Length    145.747k (_ 0.8%) i/s -    740.342k in   5.079971s
         direct-miss      1.326M (_ 1.1%) i/s -      6.633M in   5.003182s
Rack::Content-Length-miss
                        146.076k (_ 0.9%) i/s -    736.064k in   5.039301s

```

Results using above approach (implemented at https://github.com/jeremyevans/rack/tree/utils-indifferent):

```
Warming up --------------------------------------
              direct    48.653k i/100ms
Rack::Content-Length    32.365k i/100ms
         direct-miss    58.839k i/100ms
Rack::Content-Length-miss
                        32.129k i/100ms
Calculating -------------------------------------
              direct    903.400k (_ 0.9%) i/s -      4.525M in   5.008995s
Rack::Content-Length    488.650k (_ 0.5%) i/s -      2.460M in   5.033868s
         direct-miss      1.339M (_ 0.6%) i/s -      6.708M in   5.008151s
Rack::Content-Length-miss
                        484.536k (_ 0.4%) i/s -      2.442M in   5.039561s
```

So about a 3x improvement.  That may be worth missing keys with weird capitalization.  Thoughts on that approach?

**Comment by schneems** [comment]
I’m not at a computer right now so reviewing your code is not super easy.

I think my open question is if these optimizations are still measurably better or not. I’ve made a lot of improvements in the ability to measure library performance differences with

https://github.com/schneems/derailed_benchmarks/blob/master/README.md#i-made-a-patch-to-to-rails-how-can-i-tell-if-it-made-my-rails-app-faster-and-test-for-statistical-significance

I’ll put this on my backlog to see if the latest rack with my patch applied is still noticeable or not.

I think compatibility was the deal killer for my patch, which I knew when I opened it, but it is still an interesting proof of concept. If we can get faster code and preserve more compat I think it’s worth investigating, but I do want to make sure the difference is measurable from a “real world” perspective.

**Comment by ioquatix** [comment]
I think we can also avoid using `HeaderHash` in the vast majority of cases with `Rack::Response` by simply passing `@headers` through unchanged. This would avoid a lot of pointless busy work when using `Rack::Response`.

**Comment by jeremyevans** [comment]
@ioquatix do you mean lazily instantiating the `HeaderHash` instance as needed?  I don't think you can get rid of it completely and maintain backwards compatibility.

**Comment by ioquatix** [comment]
Ah, you mean because it's exposed via `#headers`?

**Comment by jeremyevans** [comment]
And related methods: `has_header?`, `get_header`, `set_header`, `delete_header`, `[]`, `[]=`

Check the specs regarding the behavior.  I looked into removing `HeaderHash` from `Response` while working on the branch I linked above, and could not find a way to make it backwards compatible.

**Comment by ioquatix** [comment]
Agreed.

Another potential optimisation: If headers are assumed to be mutable, we should just detect if `headers` is already an instance of `HeaderHash` and reuse it.

**Comment by jeremyevans** [comment]
That does seem like a good idea. We would probably want a new API for that, maybe something like `Utils::HeaderHash.convert(hash)`, or `Utils::HeaderHash(hash)` (similar to Kernel-style conversion methods).  We could then switch all middleware to use that, and then you pay the conversion cost only once instead of once per middleware.  Maybe then we don't need to worry as much regarding performance.

**Comment by ioquatix** [comment]
@jeremyevans Yeah, I've done stuff like that before. Normally I prefer something like `HeaderHash.[]` simply because it hangs off the same namespace as the class rather than a method (e.g. Rack::HeaderHash()`) in the parent module.

**Comment by schneems** [comment]
> Results using above approach (implemented at https://github.com/jeremyevans/rack/tree/utils-indifferent):


Sorry this dropped off my radar, I've queued up a test with CodeTriage on the `schneems/utils-indifferent` branch. I plan on running this tonight

Since the perf change is only one commit the derailed invocation is a little cleaner:

```
$ git co `schneems/utils-indifferent`
$ DERAILED_PATH_TO_LIBRARY="$(bundle show rack)" bundle exec derailed exec perf:library
```

If you want to set up CodeTriage locally, you can also run the test.

**Comment by schneems** [comment]
It’s worth pointing out that https://github.com/rack/rack/commit/d2e608e8d54b76ec7e52bb7c6508c5a5259c934e did not address the issue here. The code I was testing against already had the optimization of not allocating a hash if it was a HeaderHash. This is a separate concern, the HeaderHash itself is expensive.

**Comment by jeremyevans** [comment]
@schneems I'm interested in the CodeTriage results.  However, assuming we are memoizing `HeaderHash`, I think it will only be significantly faster if no `HeaderHash` is created, which means CodeTriage would have to avoid `Rack::Response` (not sure if it does or not).

**Comment by schneems** [comment]
@jeremyevans I ran a few times with 200 measurements and none were statistically significant. I then ran 1200 times to see if more samples could give us confidence. It was also not statistically significant. Here's the raw results:

[2020-02-04-20-21-1580869316-176640000.zip](https://github.com/rack/rack/files/4160546/2020-02-04-20-21-1580869316-176640000.zip)

I think in the scheme of a Rails app, this isn't a hotspot.

Here's my middleware for codetriage:

```
$ rake middleware
W, [2020-02-05T10:05:09.267810 #54036]  WARN -- Skylight: [SKYLIGHT] [4.2.0] Running Skylight in development mode. No data will be reported until you deploy your app.
(To disable this message for all local apps, run `skylight disable_dev_warning`.)
use Raven::Rack
use Rack::MiniProfiler
use ActionDispatch::HostAuthorization
use Rack::Sendfile
use ActionDispatch::Static
use ActionDispatch::Executor
use ActiveSupport::Cache::Strategy::LocalCache::Middleware
use RailsAutoscaleAgent::Middleware
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use Rack::Timeout
use ActionDispatch::RemoteIp
use Sprockets::Rails::QuietAssets
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use WebConsole::Middleware
use ActionDispatch::DebugExceptions
use ActionDispatch::ActionableExceptions
use ActionDispatch::Reloader
use ActionDispatch::Callbacks
use ActiveRecord::Migration::CheckPending
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use ActionDispatch::ContentSecurityPolicy::Middleware
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Rack::TempfileReaper
use Warden::Manager
use ScoutApm::Middleware
use ScoutApm::Instant::Middleware
use OmniAuth::Strategies::GitHub
run CodeTriage::Application.routes
```

So it looks like i'm not using that middleware, which would make sense that i'm not seeing any significance since we're basically comparing the same behavior to itself.

**Comment by jeremyevans** [comment]
@schneems Thanks for testing.  My branch had changes for Rack::ConditionalGet which looks like was in use, but that was probably it.

As even a much faster approach in a micro-benchmark does not make a measurable difference in a production application, I think it's fine to keep this issue closed.  The current implementation may be expensive, but unless we want to completely drop case insensitivity, there's probably going to be some performance hit for any approach.  BTW, do you still get a 23% performance improvement if you do `Rack::Utils::HeaderHash = Hash`?

**Comment by schneems** [comment]
> do you still get a 23% performance improvement if you do Rack::Utils::HeaderHash = Hash

I would have to queue up a test. I don't know.

**Comment by schneems** [comment]
I ran into an error just replacing HeaderHash with a Hash so I had to comment out some functional code for my test as well. I tested this commit https://github.com/schneems/rack/commit/fb22920fddaf32c81f66a7ddf07f278448d4859a

It is also not showing statistical significance. I think that the HeaderHash is only "expensive" on a really light weight app. My app example above didn't really do anything. Once you add in a database, caching, and some logic then any improvement is blown out of the water. This is why it's important to test in real-ish scenarios. 

I think this issue can stay closed. It doesn't look like a hotspot, at least not for CodeTriage. Thanks all for the work and the conversation. It's always a bummer when a perf investigation doesn't yield any usable speed ups but at least we ruled this avenue of optimization out as a large potential gain.

If you're interested in something a bit more actionable, the collection cache in Rails is pretty slow. I've got some PRs attempting to speed it up, but they're still slower than not using a cache. Also the PRs are not 100% without their risks:

- https://github.com/rails/rails/pull/36310
- https://github.com/rails/rails/pull/37301

Though FWIW I got a pretty huge speed up in CodeTriage by disabling collection caching in a  place:

- https://github.com/codetriage/codetriage/pull/1226
